### PR TITLE
clean rs by revision instead of creation timestamp in deployment controller

### LINKED ISSUE #82570: cleanupDeployment should sort ReplicaSet with revision rather than createTimeStamp
**What happened**:
when deployment controller cleans up old ReplicaSet, it sorts all old ReplicaSets by createTimeStamp and delete the oldest one, which may cause Deployment roll back to a wrong version.

```
        sort.Sort(controller.ReplicaSetsByCreationTimestamp(cleanableRSes))
	glog.V(4).Infof("Looking to cleanup old replica sets for deployment %q", deployment.Name)

	for i := int32(0); i < diff; i++ {
		rs := cleanableRSes[i]
		// Avoid delete replica set with non-zero replica counts
		if rs.Status.Replicas != 0 || *(rs.Spec.Replicas) != 0 || rs.Generation > rs.Status.ObservedGeneration || rs.DeletionTimestamp != nil {
			continue
		}
		glog.V(4).Infof("Trying to cleanup replica set %q for deployment %q", rs.Name, deployment.Name)
		if err := dc.client.ExtensionsV1beta1().ReplicaSets(rs.Namespace).Delete(rs.Name, nil); err != nil && !errors.IsNotFound(err) {
			// Return error instead of aggregating and continuing DELETEs on the theory
			// that we may be overloading the api server.
			return err
		}
	}
```

**How to reproduce it (as minimally and precisely as possible)**:
Suppose we set the revisionHistoryLimit of a Deployment to 3, and currently we have 3 Replicas: r1, r2, r3 (the createTimeStamp and Revision both are r3 > r2 > r1); if we upgrade and roll back 3 times continuously, there will be an expected case:

upgrade: 
&nbsp; ReplicaSet will be: r4, r3, r2 
&nbsp; createTimeStamp: r4 > r3 > r2
&nbsp; reversion: r4 > r3 > r2
rollback:  should rollback to v3
&nbsp; ReplicaSet will be: r4, r3, r2 
&nbsp; createTimeStamp: r4 > r3 > r2
&nbsp; reversion: r3 > r4 > r2

upgrade: 
&nbsp; ReplicaSet will be: r5, r4, r3 
&nbsp; createTimeStamp: r5 > r4 > r3 
&nbsp; reversion: r5 > r3 > r4
rollback: should rollback to v3
&nbsp; ReplicaSet will be: r4, r3, r2 
&nbsp; createTimeStamp: r5 > r4 > r3 
&nbsp; reversion: r3 > r5 > r4

upgrade: 
&nbsp; ReplicaSet will be: r6, r5, r4 
&nbsp; createTimeStamp: r6 > r5 > r4 
&nbsp; reversion: r6 > r5 > r4
rollback: should rollback to r3, but instead, we rollbacked to r5, and this may cause error.

**What you expected to happen**:
when deployment controller cleans up old ReplicaSet, it should sort old ReplicaSets by Revision

### PR DESCRIPTION
**Pull request by waynepeking348**
**What type of PR is this?**

/kind bug   

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #82570 reported by myself, and I also add some test cases
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Clean ReplicaSet by revision instead of creation timestamp in deployment controller
```

### CODE REVIEW

**Comment by waynepeking348** [comment]
would you please have a look at this fix? @enisoc @kow3ns

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -949,3 +949,18 @@ func GetDeploymentsForReplicaSet(deploymentLister appslisters.DeploymentLister,
 
 	return deployments, nil
 }
+
+// ReplicaSetsByRevision sorts a list of ReplicaSet by revision, using their creation timestamp or name as a tie breaker.
```
Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go:953]: nit: I don't see name being used as a tie breaker

Comment (waynepeking348) [reply pkg/controller/deployment/util/deployment_util.go:953]: > nit: I don't see name being used as a tie breaker

while, actually the func "controller.ReplicaSetsByCreationTimestamp" itself implements the log of using name as a tie breaker 
https://github.com/kubernetes/kubernetes/blob/c8f98589208e8c5a0d9262ebd90985acc55d2d5a/pkg/controller/controller_utils.go#L970-L980

**Comment by waynepeking348** [comment]
@janetkuo would you please have a look at this pr again?

**Review by janetkuo — APPROVED** [review]

**Comment by waynepeking348** [comment]
@janetkuo hi~could you please help to merge this pr? many thanks!

**Pull request merged**